File: python/nvstrings/tests/test_pad.py
# Copyright (c) 2018-2019, NVIDIA CORPORATION.

import nvstrings
from utils import assert_eq


def test_rjust():
    strs = nvstrings.to_device(["abc", "Def", None, "jLl"])
    got = strs.rjust(4)
    expected = [" abc", " Def", None, " jLl"]
    assert_eq(got, expected)


def test_pad():
    strs = nvstrings.to_device(
        ["hello", "there", "world", "1234", "-1234", None, "accént", ""]
    )
    got = strs.pad(5)
    expected = [
        "hello",
        "there",
        "world",
        " 1234",
        "-1234",
        None,
        "accént",
        "     ",
    ]
    assert_eq(got, expected)

    # pad right only
    got = strs.pad(7, "right")
    expected = [
        "hello  ",
        "there  ",
        "world  ",
        "1234   ",
        "-1234  ",
        None,
        "accént ",
        "       ",
    ]
    assert_eq(got, expected)

    # pad both with a specific character
    strs = nvstrings.to_device(
        ["hello", "there", "world", "1234", "-1234", None, "accént", ""]
    )
    got = strs.pad(9, "both", ".")
    expected = [
        "..hello..",
        "..there..",
        "..world..",
        "..1234...",
        "..-1234..",
        None,
        ".accént..",
        ".........",
    ]
    assert_eq(got, expected)


def test_ljust():
    strs = nvstrings.to_device(
        ["hello", "there", "world", "1234", "-1234", None, "accént", ""]
    )
    got = strs.ljust(7)
    expected = [
        "hello  ",
        "there  ",
        "world  ",
        "1234   ",
        "-1234  ",
        None,
        "accént ",
        "       ",
    ]
    assert_eq(got, expected)


def test_center():
    strs = nvstrings.to_device(
        ["hello", "there", "world", "1234", "-1234", None, "accént", ""]
    )
    got = strs.center(10, " ")
    expected = [
        "  hello   ",
        "  there   ",
        "  world   ",
        "   1234   ",
        "  -1234   ",
        None,
        "  accént  ",
        "          ",
    ]
    assert_eq(got, expected)


def test_zfill():
    strs = nvstrings.to_device(
        ["hello", "there", "world", "1234", "-1234", None, "accént", ""]
    )
    got = strs.zfill(6)
    expected = [
        "0hello",
        "0there",
        "0world",
        "001234",
        "-01234",
        None,
        "accént",
        "000000",
    ]
    assert_eq(got, expected)


def test_repeat():
    strs = nvstrings.to_device(
        ["hello", "there", "world", "1234", "-1234", None, "accént", ""]
    )
    got = strs.repeat(6)
    expected = [
        "hellohellohellohellohellohello",
        "theretheretheretheretherethere",
        "worldworldworldworldworldworld",
        "123412341234123412341234",
        "-1234-1234-1234-1234-1234-1234",
        None,
        "accéntaccéntaccéntaccéntaccéntaccént",
        "",
    ]
    assert_eq(got, expected)

File: cpp/custrings/tests/test_pad.cpp
#include <gtest/gtest.h>
#include <vector>

#include "nvstrings/NVStrings.h"

#include "./utils.h"

struct TestPad : public GdfTest{};

std::vector<const char*> hstrs{ "12345", "thesé", nullptr, "ARE THE", "tést strings", "" };

TEST_F(TestPad, Repeat)
{
    NVStrings* strs = NVStrings::create_from_array(hstrs.data(), hstrs.size());

    {
        NVStrings* got = strs->repeat(1);
        const char* expected[] = { "12345", "thesé", nullptr, "ARE THE", "tést strings", "" };
        EXPECT_TRUE( verify_strings(got,expected) );
        NVStrings::destroy(got);
    }
    {
        NVStrings* got = strs->repeat(2);
        const char* expected[] = { "1234512345", "theséthesé", nullptr, "ARE THEARE THE", "tést stringstést strings", "" };
        EXPECT_TRUE( verify_strings(got,expected) );
        NVStrings::destroy(got);
    }

    NVStrings::destroy(strs);
}

TEST_F(TestPad, Pad)
{
    NVStrings* strs = NVStrings::create_from_array(hstrs.data(), hstrs.size());

    {
        NVStrings* got = strs->ljust(10);
        const char* expected[] = { "12345     ", "thesé     ", nullptr, "ARE THE   ", "tést strings", "          " };
        EXPECT_TRUE( verify_strings(got,expected) );
        NVStrings::destroy(got);
    }
    {
        NVStrings* got = strs->rjust(7);
        const char* expected[] = { "  12345", "  thesé", nullptr, "ARE THE", "tést strings", "       " };
        EXPECT_TRUE( verify_strings(got,expected) );
        NVStrings::destroy(got);
    }
    {
        NVStrings* got = strs->center(9, "_");
        const char* expected[] = { "__12345__", "__thesé__", nullptr, "_ARE THE_", "tést strings", "_________" };
        EXPECT_TRUE( verify_strings(got,expected) );
        NVStrings::destroy(got);
    }

    NVStrings::destroy(strs);
}

TEST_F(TestPad, ZFill)
{
    NVStrings* strs = NVStrings::create_from_array(hstrs.data(), hstrs.size());
    NVStrings* got = strs->zfill(8);
    const char* expected[] = { "00012345", "000thesé", nullptr, "0ARE THE", "tést strings", "00000000" };
    EXPECT_TRUE( verify_strings(got,expected) );
    NVStrings::destroy(got);
    NVStrings::destroy(strs);
}

TEST_F(TestPad, Wrap)
{
    NVStrings* strs = NVStrings::create_from_array(hstrs.data(), hstrs.size());
    NVStrings* got = strs->wrap(3);
    const char* expected[] = { "12345", "thesé", nullptr, "ARE\nTHE", "tést\nstrings", "" };
    EXPECT_TRUE( verify_strings(got,expected) );
    NVStrings::destroy(got);
    NVStrings::destroy(strs);
}
